### PR TITLE
Client 66 wallet dex

### PR DESCRIPTION
**Pull request by tsigel**

### CODE REVIEW

File: src/modules/ui/services/EventManager.js [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+                let wasBalanceEvent = false;
+                statuses.forEach((statusData) => {
+
+                    if (statusData.status === AppEvent.statuses.PENDING) {
+                        return null;
+                    }
+
+                    needSave = true;
+                    if (BALANCE_EVENTS.indexOf(this._events[statusData.id].type) !== -1) {
+                        wasBalanceEvent = true;
```
Comment (xenohunter) [inline src/modules/ui/services/EventManager.js]: Rename to `hadBalanceEvents`.

File: src/modules/ui/services/EventManager.js [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                    }
+
+                    switch (statusData.status) {
+                        case AppEvent.statuses.ERROR:
+                            console.error(`Error event ${this._events[statusData.id]}`);
+                        case AppEvent.statuses.SUCCESS:
+                            console.log(`Event with id "${statusData.id}" finished!`);
+                            delete this._events[statusData.id];
```
Comment (xenohunter) [inline src/modules/ui/services/EventManager.js]: Add comment about missing breaks.

File: src/modules/utils/services/decorators.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+             * @param key
+             * @param descriptor
+             */
+            readonly(target, key, descriptor) {
+                const origin = descriptor.value;
+                descriptor.value = function (...args) {
+                    const result = origin.call(this, ...args);
+                    const paths = tsUtils.getPaths(result);
+                    paths.forEach((path) => {
+                        Object.freeze(tsUtils.get(result, path));
```
Comment (xenohunter) [inline src/modules/utils/services/decorators.js]: Fix it so all the parent paths are frozen.

File: src/modules/utils/services/utils.js [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+     * @name app.utils
+     */
+
+    /**
+     * @param $q
+     * @return {app.utils}
+     */
+    const factory = function ($q) {
+
+        class Moment {
```
Comment (xenohunter) [inline src/modules/utils/services/utils.js]: Move `Moment` to a separate file.

File: src/modules/wallet/modules/assets/controllers/AssetSendCtrl.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+                this.recepient = '';
+                this.polls.updateBalance = new Poll(this._getAsset.bind(this), this._setAsset.bind(this), 1000);
+                /**
+                 * @type {IFeeData}
+                 */
+                this.feeData = null;
+
+                this.observe('amount', this._onChangeAmount);
+                this.observe('amountAlias', this._onChangeAlias);
```
Comment (xenohunter) [inline src/modules/wallet/modules/assets/controllers/AssetSendCtrl.js]: Rename `amountAlias` to `amountMirror`.

File: src/modules/wallet/modules/assets/controllers/AssetSendCtrl.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+                this.ready = utils.when(Promise.all([
+                    assetsService.getAssetInfo(assetId),
+                    assetsService.getAssetInfo(aliasId),
```
Comment (xenohunter) [inline src/modules/wallet/modules/assets/controllers/AssetSendCtrl.js]: `aliaseId` → `mirrorId`

File: src/modules/wallet/modules/assets/controllers/AssetSendCtrl.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+                                    id: data.id,
+                                    amount: { ...this.asset, balance: this.amount }
+                                }
+                            });
+                            $mdDialog.hide();
+                        });
+                }
+            }
+
+            sendAll() {
```
Comment (xenohunter) [inline src/modules/wallet/modules/assets/controllers/AssetSendCtrl.js]: Rename to `fillMax`.

File: src/modules/wallet/modules/assets/controllers/AssetSendCtrl.js [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+                    this.amount = this.asset.balance;
+                }
+            }
+
+            cancel() {
+                $mdDialog.cancel();
+            }
+
+            onReadQrCode(result) {
+                this.recepient = result;
```
Comment (xenohunter) [inline src/modules/wallet/modules/assets/controllers/AssetSendCtrl.js]: Fix to `recipient`.

File: src/modules/wallet/modules/assets/services/AssetsData.js [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+     */
+    const factory = function (user, decorators, apiWorker, assetsService) {
+
+        class AssetsData {
+
+            getAssets() {
+                return user.getSetting('wallet.assets.assetList')
+                    .then((assetIds) => {
+                        return Promise.all(assetIds.map((assetId) => {
+                            return assetsService.getBalance(assetId);
```
Comment (xenohunter) [inline src/modules/wallet/modules/assets/services/AssetsData.js]: Replace with `API.Node.v2` methods.

File: vendors/qrcode.js [hunk 10]
```diff
@@ -0,0 +1,614 @@
+/**
```
Comment (xenohunter) [inline vendors/qrcode.js]: Remove this file.

File: src/modules/ui/services/EventManager.js [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+            }
+
+            /**
+             *
+             * @private
+             */
+            _initialize() {
+                this.ready = this._loadEvents()
+                    .then(() => {
+                        this.poll = new Poll(this._getStatuses.bind(this), this._applyStatuses.bind(this), 2000);
```
[the remark this answers is not in the document]
Comment (tsigel) [reply src/modules/ui/services/EventManager.js]: I make this in issue https://wavesplatform.atlassian.net/browse/CLIENT-70

File: src/modules/utils/services/utils.js (hunk 4 above)
Comment (xenohunter) [reply src/modules/utils/services/utils.js]: Moved in a different branch.

**Pull request merged**